sunscreen_tfhe/src/scratch.rs

```diff
@@ -126,102 +135,95 @@ impl Scratch {
     {
         assert_ne!(size_of::<T>(), 0);
 
-        let top = unsafe { &mut *self.top };
+        let alignment = usize::max(SIMD_ALIGN, align_of::<T>());
+        let u8_len = count * size_of::<T>();
 
-        // Push the top as far down until we hit the bottom or an allocation
-        // currently in use.
-        loop {
-            let prev = top.peek_prev();
+        let allocation = unsafe {
+            let allocation = (*self.stack).pop_back();
 
-            if let Some(x) = prev {
-                if x.is_free {
-                    top.prev().unwrap();
-                    continue;
-                }
-            }
-
-            break;
-        }
-
-        let layout = Layout::array::<T>(count).unwrap();
-        let req_len = layout.size() + layout.align();
-
-        let allocation = match top.peek_next() {
-            Some(d) => {
-                assert!(d.is_free);
-
-                // Resize the allocation if needed.
-                if d.data.len() < req_len {
-                    d.data.resize(req_len, 0u8);
-                }
+            if allocation.is_none() {
+                // If we don't have an existing allocation, make one
+                let allocation = Allocation {
+                    data: avec_rt!([alignment]| u8::default(); u8_len),
+                };
 
-                d.requested_len = count;
-                top.next().unwrap()
-            }
-            None => {
-                let data = vec![0u8; req_len];
+                let allocation = Box::new(allocation);
+                Box::into_raw(allocation)
+            } else if (*allocation.unwrap()).data.alignment() < alignment
+                || (*allocation.unwrap()).data.len() < u8_len
+            {
+                // If we found an allocation, but its size and len requirements
+                // are insufficient.
+                let allocation = allocation.unwrap();
+                let drop_box = Box::from_raw(allocation);
```

Comment:
I am guessing it doesn't make sense to look further down for an allocation that might have the right amount of space? This will just reallocate the first free thing it finds from what I am reading.

Reply:
Correct. This is less complex and also has the MRU property, which should be slightly more cache efficient.